core/src/main/java/org/mskcc/cbio/portal/servlet/QueryBuilder.java

```diff
@@ -235,28 +237,67 @@ protected void doPost(HttpServletRequest httpServletRequest,
-                    httpServletRequest.setAttribute(CANCER_STUDY_LIST, null);
-                    // multi studies
-                } else {
-                    _isVirtualStudy = true;
-                    httpServletRequest.setAttribute(CANCER_STUDY_ID, "all");
-                    httpServletRequest.setAttribute(CANCER_STUDY_LIST, cancerStudyList);
-                }
-                httpServletRequest.setAttribute(IS_VIRTUAL_STUDY, _isVirtualStudy);
-                httpServletRequest.setAttribute(CASE_SET_ID, httpServletRequest.getParameter(CASE_SET_ID));
-                if (httpServletRequest.getParameter(CASE_SET_ID).equals("-1")) {
-                    httpServletRequest.setAttribute(CASE_IDS,
-                        (httpServletRequest.getParameter(CASE_IDS)).replaceAll("\\\\n", "\n").replaceAll("\\\\t", "\t"));
-                }
+            Boolean addProfiles = false;
+            
+            //this is a common case for most of the requests from study view when the there is only one study
+            if (cancerStudyId != null && cancerStudyId.equals("all") && cancerStudyList != null && cancerStudyList.split(",").length == 1) {
+            		cancerStudyId = cancerStudyList.split(",")[0];
+            		addProfiles = true;
+            }
+
+            //redirect requests with cancerStudyId="all" and cancerStudyList="all" to home page
+            // handle different possible scenarios in https://github.com/cBioPortal/cbioportal/issues/3431
+            if (!(cancerStudyId != null 
+            		&& cancerStudyId.equals("all") 
+            		&& cancerStudyList != null 
+            		&& cancerStudyList.equals("all") )) {
+            	
+	            if (cancerStudyId != null || cancerStudyList != null) {
+			    		CancerStudy cancerStudy = null;
+			    		if(cancerStudyId != null) {
+			    			cancerStudy = DaoCancerStudy.getCancerStudyByStableId(cancerStudyId);
+			    		}
+			        
+			    		//single regular study
+			        if(cancerStudy != null) {
+			        		httpServletRequest.setAttribute(CANCER_STUDY_ID, cancerStudy.getCancerStudyStableId());
+			        		httpServletRequest.setAttribute(CANCER_STUDY_LIST, null);
+			        		if(addProfiles) {
+			        			cancerStudy.getGeneticProfiles();
+			        			geneticProfileIdSet = new HashSet<String>();
+			        			GeneticProfile cnaProfile      = cancerStudy.getCopyNumberAlterationProfile(true);
+			        			GeneticProfile mutationProfile =  cancerStudy.getMutationProfile();
+			        			if(cnaProfile != null) {
+			        				geneticProfileIdSet.add(cnaProfile.getStableId());
+			        			}
+			        			if(mutationProfile != null) {
+			        				geneticProfileIdSet.add(mutationProfile.getStableId());
+			        			}
+			        			 httpServletRequest.setAttribute(GENETIC_PROFILE_IDS, geneticProfileIdSet);
+			        		}
+			        } 
+			        //multiple and virtual study
+			        else {
+			        		_isVirtualStudy = true;
```

Comment:
Still curious that, if the cancerStudyId is all, but there are multiple studies in CANCER_STUDY_LIST, why would it have to be virtual study?

Reply:
this is an signal to front-end that its a multiple or virtual study

Reply:
Yeah, but isn't it supposed to be placed after line 282 `if (cancerStudyId != null && !cancerStudyId.equals("all")) {` ?

Reply:
even for the else block its considered as virtual study